File: pa1/Shuffle.java
/*Yugraj Sing
 * 
 * Main shuffle class that performs the permutations on the list
 * 
 */

import java.io.*;
import java.util.Scanner;

class Shuffle{
  public static void main(String args[]) throws IOException{
    Scanner input = null;
    PrintWriter out = null;
    String line = null;
    List[] P;
    List[] L;
    int n=0;
    
    /*exits the program if the command line args arn't met*/
    if(args.length < 2){
      System.err.println("Usage: Shuffle infile outfile");
      System.exit(1);
    }
    //initialising input and out
    input = new Scanner(new File(args[0]));
    out = new PrintWriter(new FileWriter(args[1]));
    
    input.useDelimiter("\n");
    n = input.nextInt();//get the number of permutations P
    //creates the list array 
    P = new List[n];
    L = new List[n];
    //initialize the array List
    for(int i=0; i<n; i++){
      P[i] = new List();
      L[i] = new List();
    }
    
    //insert the values from the file to the List
    for (int i=0; i<n;i++){
      line = input.next();
      insertPer(P[i],line);
      initializeList(L[i],P[i].getLength()); 
    }
    
    //main loop
    for(int i =0; i<n; i++){
      int order = 1; 
      List originalList = L[i].copy(); 
      shuffle(L[i],P[i]);
      out.print(L[i]);
      //shuffles the list L unit it back to the original L
      while( !originalList.equals(L[i])){
        shuffle(L[i],P[i]);
        order++;
      }
      out.println( " order = " + order);
    }
    
    input.close();
    out.close();    
  }
  
  /* Shuffle
   * Pre: two arguments ( Lists)
   * Pos: performs the permutation on the list L
   */ 
  public static void shuffle(List L, List P){
    int index =2;
    P.moveTo(0);
    while( P.offEnd() != true){
      L.moveTo(L.getLength()+P.getCurrent()-index);
      L.insertAfterCurrent(L.getFront());
      L.deleteFront();
      index++;
      P.moveNext(); 
    }
  }
  
  /* insertPer()
   * Inserts the values from the input file into the 
   * List
   */ 
  public static void insertPer(List P,String Line){
    int n;
    Scanner in = new Scanner(Line);
    while(in.hasNextInt()){
      P.insertBack(in.nextInt());
    }
  }
  
  /* initializeList()
   * initializes a List L with values 1...P.length
   */ 
  public static void initializeList(List L, int temp){
    for(int i = 1; i<=temp; i++){
      L.insertBack(i);
    }
  }
  
}

File: pa2/Shuffle.c
/*Yugraj Singh
 *Shuffle.c
 *Main program that performs the permutations
 */

#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include "List.h"

#define MAX_LEN 300
/*shuffle
 *Pre: takes to ListRef
 *Pos: shuffles the list L by the permutations listed in List P
 */ 
void shuffle(ListRef L, ListRef P){
	int index = 2;
	int holdIndex;
	moveTo(P,0);
	while( offEnd(P) != 1){
		holdIndex = getLength(L) + getCurrent(P)-index;
		moveTo(L,holdIndex);
		insertAfterCurrent(L,getFront(L));
		deleteFront(L);
		index++;
		moveNext(P);
	}
}


int main(int argc, char* argv[]){
	FILE* in;
	FILE* out;
	char line[MAX_LEN];
	char* token;
	int order, size;
	
	
	if (argc !=3){
		printf("Usage: %s infile outfile\n",argv[0]);
		exit(EXIT_FAILURE);
	}
	/*opens the input and output files*/
	in = fopen(argv[1], "r");
	out = fopen(argv[2], "w");
        /* check to make sure the input and output files are functioning*/
	if ( in == NULL){
		printf("Unable to open file %s for reading\n",argv[1]);
		exit(EXIT_FAILURE);
	}
	if ( out == NULL){
		printf("Unable to open file %s for writing\n",argv[2]);
		exit(EXIT_FAILURE);
	}
	
	/*reads the first line in file and assigns the value to size*/ 
	if( fgets(line,MAX_LEN,in) != NULL){
		token = strtok(line," \n");
		while(token!= NULL){
			size = atoi(token);
			token = strtok(NULL, " \n");
		}
	}
	
	/* main loop*/
	int i;
	for( i=0; i< size; i++){
		fgets(line,MAX_LEN,in);
		token = strtok(line," \n");

		ListRef P = newList();
		ListRef L = newList();
		
		/*inserts the data values from input file*/
		while(token!=NULL){
			insertBack(P,atoi(token));
			token = strtok(NULL," \n");
		}

		/* creates a List to perform the permutations on*/
		int j = 0;
		while( j<getLength(P)){
			insertBack(L,j+1);
		j++;
		}
		
		/*copys the list*/
		ListRef listCopy = copyList(L);
		shuffle(L,P);
		printList(out,L);
		order = 1;
		
		/*counts the number of permutations being performed*/
		while( !equals(listCopy,L)){
			shuffle(L,P);
			order++;
		}
	
		/*print the order, and frees the Heap memory from the list*/
		fprintf(out," order = %d\n", order);
		freeList(&listCopy);
		freeList(&L);
		freeList(&P);		
	}
	
	
	
       /* closes the input and output files*/
	fclose(in);
	fclose(out);
	
	return(0);
}
